lib/iris/experimental/ugrid/utils.py

```diff
@@ -0,0 +1,321 @@
+                indexcoord_metadata = sub_metadata
+            elif sub_metadata != indexcoord_metadata:
+                # Compare subsequent occurrences (from other region-cubes)
+                err = (
+                    f"{sub_str} has an index coord "
+                    f'"{index_coord_name}" whose ".metadata" does not '
+                    f"match that of the first region-cube :  "
+                    f"{sub_metadata} != {indexcoord_metadata}."
+                )
+
+    # Use the mesh_dim to transpose inputs + outputs, if required, as it is
+    # simpler for all the array operations to always have the mesh dim *last*.
+    if mesh_dim == mesh_cube.ndim - 1:
+        # Mesh dim is already the last one : no tranpose required
+        untranspose_dims = None
+    else:
+        dim_range = np.arange(mesh_cube.ndim, dtype=int)
+        # Transpose all inputs to mesh-last order
+        tranpose_dims = [i_dim for i_dim in dim_range if i_dim != mesh_dim] + [
+            mesh_dim
+        ]  # chop out mesh_dim + put it at the end
+
+        def transposed_copy(cube, dim_order):
+            cube = cube.copy()
+            cube.transpose(dim_order)
+            return cube
+
+        mesh_cube = transposed_copy(mesh_cube, tranpose_dims)
+        submesh_cubes = [
+            transposed_copy(region_cube, tranpose_dims)
+            for region_cube in submesh_cubes
+        ]
+
+        # Also prepare for transforming the output back to the original order
+        untranspose_dims = dim_range.copy()
+        # Neat trick to produce the reverse operation
+        untranspose_dims[tranpose_dims] = dim_range
+
+    #
+    # Here's the core operation..
+    #
+    def fill_region(target, regiondata, regioninds):
+        if not target.flags.writeable:
+            # The initial input can be a section of a da.zeros(), which has no
+            # real array "behind" it.  This means that real arrays created in
+            # memory are only chunk-sized, but it also means that 'target' may
+            # not be writeable.  So take a copy to fix that, where needed.
+            target = target.copy()
+        # N.B. Indices are basically 1D, but may have leading *1 dims for
+        # alignment, to satisfy da.map_blocks
+        assert all(size == 1 for size in regioninds.shape[:-1])
+        inds = regioninds.flatten()
+        # Assign blocks with indexing on the last dim only
+        target[..., inds] = regiondata
+        return target
+
+    # Create an initially 'empty' (all-masked) dask array matching the input.
+    # N.B. this does not use the mesh_cube.lazy_data() array, but only its
+    # shape and dtype, since the data itself is not used in the calculation.
+    # N.B. chunking matches the input cube, allowing performance control.
+    input_data = mesh_cube.lazy_data()
+    result_array = da.ma.masked_array(
+        da.zeros(
+            input_data.shape,
+            dtype=input_data.dtype,
```

Comment:
NOTE: just spotted, this should be "result_dtype".
To be fixed in forthcoming commits, with added testcases

Reply:
@pp-mo I don't think this matters :thinking: 

At this point we know that all the `dtypes` are equal between the `mesh_cube` and the `submesh_cubes`, right?

Reply:
No, that's not required.
The submesh data has it's own phenomenon type, e.g. name + units,  which could be different from 'mesh_cube' as that's only a template (logically, just for the mesh itself, but we fudged that point).
So if it  has its own name+units, it makes sense to include dtype with that -- e.g. it could be a category number of a classification, or a COUNT statistic

Reply:
So.. I just added a test where the submesh_cube info is all ints.